# <font color=red> Exploration 2. Iris(붓꽃) 분류</font>

프랑스의 국화인 __Iris(붓꽃)__ 데이터를 분류해보자.

Scikit-learn 에서는 예제로 활용해 볼 수 있는 7가지의 Toy datasets 와 7가지의 Real world datasets 를  제공한다.  

Toy datasats 중 한가지가 이번에 다뤄볼 __Iris dataset__ 이다.  

iris dataset 의 sepal(꽃받침)과 petal(꽃잎)의 길이와 폭을 가지고 세 개의 품종(setosa, versicolor, virginica) 으로 분류하는 모델을 만들어보자!

![image](https://user-images.githubusercontent.com/84179578/125756842-ae86e9e6-1ae4-42c5-a3cf-90bda928748c.png)

## 데이터 정보 확인

데이터셋을 다루기 전에 먼저 데이터셋의 정보를 확인해야한다.  
데이터에 대한 이해도는  해당 데이터를 활용한 결과와 성능에 큰 영향을 미치기때문!

[Scikit-learn Iris plants dataset](https://scikit-learn.org/stable/datasets/toy_dataset.html#iris-plants-dataset) 에서 iris 데이터셋에 대한 정보를 확인해보자.

![image](https://user-images.githubusercontent.com/84179578/125756953-1a934b1a-3db5-4589-a51d-27ba791cc13a.png)

위와 같이, 데이터에 대한 다양한 기본 정보들을 알 수 있다.

## 데이터 불러오기

In [1]:
from sklearn.datasets import load_iris

iris = load_iris()

print(dir(iris))  # dir : 객체가 어떤 변수와 메소드를 가지고 있는지 나열

['DESCR', 'data', 'feature_names', 'filename', 'frame', 'target', 'target_names']


In [2]:
iris.keys()  # iris 에 담긴 정보 확인

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])

In [3]:
iris_data = iris.data   # iris 의 데이터를 변수에 저장

print(iris_data.shape)
print(iris_data[0])

(150, 4)
[5.1 3.5 1.4 0.2]


iris의 데이터는 총 150개의 데이터가 각각 4개의 정보를 담고 있는 것을 확인 할 수 있다.  

0번 index의 데이터를 보니, 순서대로 sepal length, sepal width, petal length, petal width 의 값이 담겨있다.

그럼 이제 위 4개의 정보를 입력했을때 붓꽃의 품종을 출력하도록 해야한다.

머신러닝 모델이 출력해야 하는 정답을 __라벨(label)__ 또는 __타겟(target)__ 이라고 한다.  
iris 데이터에서 `target`에 해당 타겟 정보가 담겨있다.

In [4]:
iris_label = iris.target   # 타겟정보를 변수에 저장
print(iris_label.shape)
iris_label

(150,)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

이외의 메소드들의 역할은 다음과 같다.
- `target_names` : 라벨 이름
- `DESCR` : 데이터셋에 대한 설명
- `feature_names` : 각 feature 이름
- `filename` : 데이터셋 파일이 저장된 경로

In [5]:
iris.target_names   # 라벨 이름

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

순서 그대로 target 값이 
`0` 이면 `setosa`,
`1` 이면 `versicolor`,
`2` 이면 `virginica` 를 나타낸다

In [6]:
print(iris.DESCR)    # 데이터셋에 대한 설명

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

    :Number of Instances: 150 (50 in each of three classes)
    :Number of Attributes: 4 numeric, predictive attributes and the class
    :Attribute Information:
        - sepal length in cm
        - sepal width in cm
        - petal length in cm
        - petal width in cm
        - class:
                - Iris-Setosa
                - Iris-Versicolour
                - Iris-Virginica
                
    :Summary Statistics:

    ============== ==== ==== ======= ===== ====================
                    Min  Max   Mean    SD   Class Correlation
    ============== ==== ==== ======= ===== ====================
    sepal length:   4.3  7.9   5.84   0.83    0.7826
    sepal width:    2.0  4.4   3.05   0.43   -0.4194
    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
    ============== ==== ==== ======= ===== ===========

----

## 데이터 다루기 (training set, test set 나누기)

iris 데이터는 행과 열이 있는 2차원 데이터이므로 2차원 배열 데이터를 다루는데 가장 많이 쓰이는 도구인 `pandas` 를 활용하겠다.  

먼저, iris 데이터를 DataFrame 자료형으로 변환하자.

In [7]:
import pandas as pd

iris_df = pd.DataFrame(data=iris_data, columns=iris.feature_names)
iris_df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

타겟 데이터도  label 이라는 컬럼으로 추가한다.

In [8]:
iris_df["label"] = iris.target
iris_df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     label  
0        0  
1        0  
2        0  
3        0  
4        0  
..     ...  
145      2  
146      2  
147      2  
148      2  
149      2  

[150 rows x 5 columns]

`sklearn,model_selection` 패키지의 `train_test_split` 함수를 활용하여, training set 와 test set 을 나눌수 있다.

In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(iris_data, 
                                                    iris_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

print('X_train 개수: ', len(X_train), ', X_test 개수: ', len(X_test))

X_train 개수:  120 , X_test 개수:  30


`train_test_split` 함수의 파라미터를 정리하면
- __첫번째 파라미터__ : feature (특징)
- __두번째 파라미터__ : label (정답)
- __test_size__ : training set 과 test set 의 비율.  
위 코드에서는 전체 데이터의 20%를 test set 으로 사용하여, 8:2 비율로 training set 과 test set 으로 나눔
- __random_state__ :  랜덤하게 섞는 과정에서 일정한 결과값을 보여주기 위해 지정

In [10]:
X_train.shape, y_train.shape

((120, 4), (120,))

In [11]:
X_test.shape, y_test.shape

((30, 4), (30,))

## Decision Tree 모델 사용하여 학습시키기

지금 하고자하는 iris 데이터에 대한 문제는 label(정답) 이 있으므로 __지도학습 (Supervised Learning)__ 에 해당한다.  
또한 입력 데이터를 통해 특정 카테고리 중 하나로 분류해내는 __분류 (Classification)__ 에 해당한다.

수많은 분류 모델 중 __Decision Tree 모델__ 을 사용해보겠다.

Decision Tree (결정트리) 에 대한 내용은 블로그에 자세히 정리하였다.  
> 블로그 링크 추가 예정

Decision Tree 는 `sklearn.tree` 패키지 안에 `DecisionTreeClassifier` 라는 이름으로 내장되어있다.

In [12]:
from sklearn.tree import DecisionTreeClassifier

decision_tree = DecisionTreeClassifier(random_state=32)
print(decision_tree._estimator_type)

classifier


In [13]:
decision_tree.fit(X_train, y_train)

DecisionTreeClassifier(random_state=32)

`fit` 함수를 통해 모델을 학습시켰다. 모델을 training set 으로 학습시킨다는 개념이 training set에 맞게 모델을 맞추는것(fit) 이라고 할 수 있다.

In [15]:
y_pred = decision_tree.predict(X_test)  # X_test 데이터로 예측한 값을 y_pred 에 저장
y_pred

array([2, 1, 0, 1, 2, 0, 1, 1, 0, 1, 2, 1, 0, 2, 0, 2, 2, 2, 0, 0, 1, 2,
       1, 1, 2, 2, 1, 1, 2, 2])

In [16]:
y_test

array([2, 1, 0, 1, 2, 0, 1, 1, 0, 1, 1, 1, 0, 2, 0, 1, 2, 2, 0, 0, 1, 2,
       1, 2, 2, 2, 1, 1, 2, 2])

`y_pred` 와 `y_test` 데이터를 눈으로 간단히 비교해보면 어느 정도 잘 맞은것을 확인할 수 있다.  

`sklearn.metrics` 패키지를 이용하면 예측한 결과에 대한 수치를 편리하게 확인할 수 있다.  
성능을 평가하는 방법에는 다양한 척도가 있는데, 그 중 __정확도 (Accuracy)__ 를 확인해보자.  

정확도는 전체 개수 중 맞은 것의 개수의 수치이다.


In [17]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred) # 실제 라벨값, 예측한 라벨값
accuracy

0.9

Iris 데이터를 이용해 Decision Tree 모델을 학습시키고 분류하는 전체 과정은 다음과 같다.

In [18]:
# (1) 필요한 모듈 import
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report

# (2) 데이터 준비
iris = load_iris()
iris_data = iris.data
iris_label = iris.target

# (3) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(iris_data, 
                                                    iris_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

# (4) 모델 학습 및 예측
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00         7
           1       0.91      0.83      0.87        12
           2       0.83      0.91      0.87        11

    accuracy                           0.90        30
   macro avg       0.91      0.91      0.91        30
weighted avg       0.90      0.90      0.90        30



여기서 모델을 바꾸고 싶다면 `(4) 모델 학습 및 예측` 부분에서 사용할 모델로 바꾸면 된다.  

다른 scikit-learn 내장 분류 모델에 대한 자세한 내용은 블로그에 정리하겠다.
> 블로그 링크 추가 예정

## RandomForest 모델 사용하여 학습시키기

간단히, Decision Tree를 여러개 모아놓은 __RandomForest__ 모델을 사용하겠다.  
단일 모델을 여러개 사용함으로써 한 개만 사용할 때 생기는 문제를 극복하는 기법을 __앙상블(Ensemble) 기법__ 이라고 한다.

`sklearn.ensemble` 패키지 내에 있는 `RandomForestClssifier` 을 사용해 RandomForest 모델을 사용해보자. 

In [19]:
from sklearn.ensemble import RandomForestClassifier

X_train, X_test, y_train, y_test = train_test_split(iris_data, 
                                                    iris_label, 
                                                    test_size=0.2, 
                                                    random_state=21)

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        11
           1       1.00      0.83      0.91        12
           2       0.78      1.00      0.88         7

    accuracy                           0.93        30
   macro avg       0.93      0.94      0.93        30
weighted avg       0.95      0.93      0.93        30



## Support Vector Machine (SVM) 모델 사용하여 학습시키기

In [20]:
from sklearn import svm
svm_model = svm.SVC()

X_train, X_test, y_train, y_test = train_test_split(iris_data, 
                                                    iris_label, 
                                                    test_size=0.2, 
                                                    random_state=21)

X_train, X_test, y_train, y_test = train_test_split(iris_data, 
                                                    iris_label, 
                                                    test_size=0.2, 
                                                    random_state=21)


svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        11
           1       0.91      0.83      0.87        12
           2       0.75      0.86      0.80         7

    accuracy                           0.90        30
   macro avg       0.89      0.90      0.89        30
weighted avg       0.91      0.90      0.90        30



## Stochastic Gradient Descent Classifier (SGDClassifier) 모델 사용하여 학습시키기

In [21]:
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()

X_train, X_test, y_train, y_test = train_test_split(iris_data, 
                                                    iris_label, 
                                                    test_size=0.2, 
                                                    random_state=21)

X_train, X_test, y_train, y_test = train_test_split(iris_data, 
                                                    iris_label, 
                                                    test_size=0.2, 
                                                    random_state=21)


sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        11
           1       0.86      1.00      0.92        12
           2       1.00      0.71      0.83         7

    accuracy                           0.93        30
   macro avg       0.95      0.90      0.92        30
weighted avg       0.94      0.93      0.93        30



## 모델 성능 평가의 다양한 척도

이때까지 우리는 정확도로 모델의 성능을 평가하였다. 하지만 정확도는 확실한 단점이 있다. 

따라서 __오차행렬 (Confusion)__ 을 사용하여 다양한 척도로 모델의 성능을 평가한다.  
오차 행렬은 `sklearn.metrics` 패키지 내의 `confusion_matrix`로 확인할 수 있다.

정확도의 단점과 오차행렬에 대한 자세한 내용은 블로그에 정리하였다.
> 블로그 링크 추가 예정